docs/datamodel/build-a-datamodel.rst

```diff
@@ -0,0 +1,627 @@
+    |   |       `-- record-v1.1.0.json
+    |   |-- mappings
+    |   |   |-- __init__.py
+    |   |   `-- v6
+    |   |       |-- __init__.py
+    |   |       `-- records
+    |   |           `-- record-v1.0.0.json
+    |   |           `-- record-v1.1.0.json
+
+
+This allows you to simultaneously store old and new records - i.e. you don't
+have to take down your service for hours to migrate you millions of records
+from one version to a new one.
+
+Now of course, old records will be sent to the ``records-record-v1.0.0`` index
+and new records will be sent to the ``records-record-v1.1.0`` index. However,
+a special Elasticsearch *index alias* ``records`` is also created, that allows
+you to search over both old and new records, thus smoothly handling data model
+evolution.
+
+**Discovery of mappings**
+
+Invenio is using standard Python entry points to discover your data model
+package's Elasticsearch mappings. Thus, you'll see in the ``setup.py`` an entry
+point group ``invenio_search.mappings``:
+
+.. code-block:: python
+
+    setup(
+        # ...
+        entry_points={
+            'invenio_search.mappings': [
+                'records = my_datamodel.mappings'
+            ],
+            # ...
+        },
+    )
+
+Note, that the left-hand-side of the entry point
+``records = my_datamodel.mappings``, defines the folder name/index alias (i.e.
+``records`` and that the right-hand-side defines the Python import path to the
+``mappings`` package.
+
+.. note::
+
+    A typical mistake is to forget to add a blank ``__init__.py`` file inside
+    the ``mappings``, ``v5`` and ``v6`` folders, in which case the entry points won't be correctly discovered.
+
+Define a Marshmallow schema
+---------------------------
+`Marhsmallow <https://marshmallow.readthedocs.io/en/3.0/index.html>`_ is a
+Python library that helps you write highly advanced
+serialization/deserialization/validation rules for your input/output data.
+
+Marshmallow use in Invenio is optional, but is usually very helpful when you go
+beyond purely structural data validation - e.g. validating one field given the
+value of another field.
+
+In Invenio the Marshmallow schemas are located in the ``marshmallow`` Python
+module. You may have multiple Marshmallow schemas depending on your
+serialization and deserialization needs.
+
+.. code-block:: shell
+
+    |-- my_datamodel
+    |   |-- marshmallow
+    |   |   |-- __init__.py
+    |   |   `-- json.py
+
+Below is a simplified example of a Marshmallow schema you could use in
+``json.py`` (note, the scaffolded data model package, includes a more complete
+example):
+
+.. code-block:: python
+
+    from invenio_records_rest.schemas import StrictKeysMixin
+    from marshmallow import fields
+
+    class RecordSchemaV1(StrictKeysMixin):
+        metadata = fields.Raw()
+        created = fields.Str()
+        revision = fields.Integer()
+        updated = fields.Str()
+        links = fields.Dict()
+        id = fields.Str()
```

Comment:
Would it be too much to add here something characteristic of Marshmallow? For instance:
```diff
+def get_id(obj, context):
+   """Get record id."""
+   pid = context.get('pid')
+   return pid.pid_value if pid else missing
+
class RecordSchemaV1(StrictKeysMixin):
    """Record schema."""

    metadata = fields.Nested(MetadataSchemaV1)
    created = fields.Str(dump_only=True)
    revision = fields.Integer(dump_only=True)
    updated = fields.Str(dump_only=True)
    links = fields.Dict(dump_only=True)
-   id = fields.Str()
+   id = fields.Function(
+      serialize=get_id,
+      deserialize=get_id)
```

It doesn't have to be this, but something that paves the way for the next chapter explanation, so you feel like there are some differences (as the one explained in the [Elasticsearch mappings with the stemming example](https://github.com/inveniosoftware/invenio/pull/3889/files#diff-d7b881a94984fe8ca37d89579c49fc0dR195)).

Reply:
We might be able to update it with the new persistent identifier field.....the reason for not using the exact version from the generated skeleton was to not make it too compelx.